src/queue.rs

```diff
@@ -248,73 +222,187 @@ pub fn pop<'d, S: MultiwriteNorFlash>(
     flash_range: Range<u32>,
     data_buffer: &'d mut [u8],
 ) -> Result<Option<&'d mut [u8]>, Error<S::Error>> {
-    assert_eq!(flash_range.start % S::ERASE_SIZE as u32, 0);
-    assert_eq!(flash_range.end % S::ERASE_SIZE as u32, 0);
+    pop_many(flash, flash_range).next(data_buffer)
+}
 
-    assert!(S::ERASE_SIZE >= S::WORD_SIZE * 4);
-    assert!(S::WORD_SIZE <= MAX_WORD_SIZE);
+/// Iterator for pop'ing elements in the queue.
+pub struct PopIterator<'d, S: MultiwriteNorFlash> {
+    iter: QueueIterator<'d, S>,
+}
+
+impl<'d, S: MultiwriteNorFlash> PopIterator<'d, S> {
+    /// Pop the next event.
+    ///
+    /// The data is written to the given `data_buffer`` and the part that was written is returned.
+    /// It is valid to only use the length of the returned slice and use the original `data_buffer`.
+    /// The `data_buffer` may contain extra data on ranges after the returned slice.
+    /// You should not depend on that data.
+    ///
+    /// If the data buffer is not big enough an error is returned.
+    pub fn next<'m>(
+        &mut self,
+        data_buffer: &'m mut [u8],
+    ) -> Result<Option<&'m mut [u8]>, Error<S::Error>> {
+        if let Some((item, item_address)) = self.iter.next(data_buffer)? {
+            let (header, data_buffer) = item.destruct();
+            let ret = &mut data_buffer[..header.length as usize];
+            self.iter.erase(header, item_address)?;
+
+            Ok(Some(ret))
+        } else {
+            Ok(None)
+        }
+    }
+}
 
-    let oldest_page = find_oldest_page(flash, flash_range.clone())?;
+/// Iterator for peek'ing elements in the queue.
+pub struct PeekIterator<'d, S: NorFlash> {
+    iter: QueueIterator<'d, S>,
+}
+
+impl<'d, S: NorFlash> PeekIterator<'d, S> {
+    /// Peek at the next event.
+    ///
+    /// The data is written to the given `data_buffer`` and the part that was written is returned.
+    /// It is valid to only use the length of the returned slice and use the original `data_buffer`.
+    /// The `data_buffer` may contain extra data on ranges after the returned slice.
+    /// You should not depend on that data.
+    ///
+    /// If the data buffer is not big enough an error is returned.
+    pub fn next<'m>(
+        &mut self,
+        data_buffer: &'m mut [u8],
+    ) -> Result<Option<&'m mut [u8]>, Error<S::Error>> {
+        Ok(self.iter.next(data_buffer)?.map(|(item, _)| {
+            let (header, data_buffer) = item.destruct();
+            &mut data_buffer[..header.length as usize]
+        }))
+    }
+}
+
+/// An iterator-like interface for peeking into events stored in flash.
+pub struct QueueIterator<'d, S: NorFlash> {
```

Comment:
I don't think this needs to be public
```suggestion
struct QueueIterator<'d, S: NorFlash> {
```